libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1707,19 +1708,52 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 				if gt(length, 0xffffffffffffffff) { <panic>() }
 				<resizeArray>(dst, length)
 
-				let srcPtr := <srcDataLocation>(src)
-
-				let dstPtr := <dstDataLocation>(dst)
+				let srcSlot := <srcDataLocation>(src)
+				let dstSlot := <dstDataLocation>(dst)
 
 				let fullSlots := div(length, <itemsPerSlot>)
-				let i := 0
-				for { } lt(i, fullSlots) { i := add(i, 1) } {
-					sstore(add(dstPtr, i), <maskFull>(sload(add(srcPtr, i))))
-				}
-				let spill := sub(length, mul(i, <itemsPerSlot>))
-				if gt(spill, 0) {
-					sstore(add(dstPtr, i), <maskBytes>(sload(add(srcPtr, i)), mul(spill, <bytesPerItem>)))
-				}
+				let spill := sub(length, mul(fullSlots, <itemsPerSlot>))
+
+				<?sameStride>
+					let i := 0
+					for { } lt(i, fullSlots) { i := add(i, 1) } {
+						sstore(add(dstSlot, i), <maskFull>(sload(add(srcSlot, i))))
+					}
+					if gt(spill, 0) {
+						sstore(add(dstSlot, i), <maskBytes>(sload(add(srcSlot, i)), mul(spill, <dstStride>)))
+					}
+				<!sameStride>
+					let srcSlotValue := sload(srcSlot)
```

Comment:
Can't this move inside the loop?

Reply:
Loop is done over slots in dst. So it lives independently of the loop. This is important for cases when copying, for example `bytes8[]` to `bytes10`. So the counters are not reset at the same time...